certora/specs/VaultERC4626.spec

```diff
@@ -0,0 +1,535 @@
+    to_mathint(totalSupply(e)) == sumOfBalances;
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+////                    #     State Transition                             /////
+////////////////////////////////////////////////////////////////////////////////
+
+//run: https://prover.certora.com/output/65266/3ef25c98a7e34422bcf177d853662b5f/?anonymousKey=ca43e967a607a404f34b39c70f6517e90dac0902
+rule totalsMonotonicity() {
+    method f; env e; calldataarg args;
+    require e.msg.sender != currentContract; 
+    uint256 totalSupplyBefore = totalSupply(e);
+    uint256 totalAssetsBefore = totalAssets(e);
+    address receiver;
+    safeAssumptions(e, receiver, e.msg.sender);
+    callReceiverFunctions(f, e, receiver);
+
+    uint256 totalSupplyAfter = totalSupply(e);
+    uint256 totalAssetsAfter = totalAssets(e);
+    
+    // possibly assert totalSupply and totalAssets must not change in opposite directions
+    assert totalSupplyBefore < totalSupplyAfter  <=> totalAssetsBefore < totalAssetsAfter,
+        "if totalSupply changes by a larger amount, the corresponding change in totalAssets must remain the same or grow";
+    assert totalSupplyAfter == totalSupplyBefore => totalAssetsBefore == totalAssetsAfter,
+        "equal size changes to totalSupply must yield equal size changes to totalAssets";
+}
+
+rule underlyingCannotChange() {
+    address originalAsset = asset();
+
+    method f; env e; calldataarg args;
+    f(e, args);
+
+    address newAsset = asset();
+
+    assert originalAsset == newAsset,
+        "the underlying asset of a contract must not change";
+}
+
+////////////////////////////////////////////////////////////////////////////////
+////                    #   High Level                                    /////
+////////////////////////////////////////////////////////////////////////////////
+
+// passing
+// run: https://prover.certora.com/output/65266/1912c053cdf8485087f2c050146c64aa/?anonymousKey=a12e3d573258a4d8136a19b612448a50f80b9a21
+rule dustFavorsTheHouse(uint assetsIn )
+{
+    env e;
+        
+    require e.msg.sender != currentContract;
+    safeAssumptions(e,e.msg.sender,e.msg.sender);
+    uint256 totalSupplyBefore = totalSupply(e);
+
+    uint balanceBefore = currentContract.balanceOf(e, currentContract);
+
+    uint shares = deposit(e,assetsIn, e.msg.sender);
+    uint assetsOut = redeem(e,shares,e.msg.sender,e.msg.sender);
+
+    uint balanceAfter = currentContract.balanceOf(e, currentContract);
+    assert balanceAfter >= balanceBefore;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+////                       #   Risk Analysis                           /////////
+////////////////////////////////////////////////////////////////////////////////
+
+
+invariant vaultSolvency(env e)
+    totalAssets(e) >= totalSupply(e)  && userAssets(e, currentContract) >= require_uint256(cache_cash(e))  {
```

Comment:
1. No I thought this at one point as well, but I think we want to check the relation between the absolute values. `convertToAssets` essentially scales the shares by the ratio of totalAssets to totalShares. See:
```
    function toAssetsDown(Shares amount, VaultCache memory vaultCache) internal pure returns (Assets) {
        (uint256 totalAssets, uint256 totalShares) = ConversionHelpers.conversionTotals(vaultCache);
        unchecked {
            return TypesLib.toAssets(amount.toUint() * totalAssets / totalShares);
        }
    }
```

We want to check that the actual underlying asset total is greater than the shares (not scaled by this ratio)

Reply:
2. Originally in this rule I had `userAssets(e, currentContract) >= totalAssets(e)` which is what I think you are looking for, but I changed this to `cache_cash (which returns `vaultCache.cash`).

Note that `totalAssets = vaultCache.cash + vaultCache.totalBorrows`: 


    function totalAssetsInternal(VaultCache memory vaultCache) internal pure virtual returns (uint256) {
        // total assets can exceed Assets max amount (MAX_SANE_AMOUNT)
        return vaultCache.cash.toUint() + vaultCache.totalBorrows.toAssetsUp().toUint();
    }
    
When users move money in and out of the vault the cash balance is what's changed rather than borrows.